sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java

```diff
@@ -0,0 +1,501 @@
+          ValueProvider.StaticValueProvider.of(accessKey),
+          ValueProvider.StaticValueProvider.of(secretKey));
+    }
+
+    public static DynamoDBConfiguration create(
+        ValueProvider<String> endpointUrl,
+        ValueProvider<String> region,
+        ValueProvider<String> accessKey,
+        ValueProvider<String> secretKey) {
+      checkArgument(region != null, "region can not be null");
+      checkArgument(accessKey != null, "accessKey can not be null");
+      checkArgument(secretKey != null, "secretKey can not be null");
+      return new AutoValue_DynamoDBIO_DynamoDBConfiguration.Builder()
+          .setEndpointUrl(endpointUrl)
+          .setRegion(region)
+          .setAwsAccessKey(accessKey)
+          .setAwsSecretKey(secretKey)
+          .build();
+    }
+
+    public DynamoDBConfiguration withRegion(String region) {
+      return withRegion(ValueProvider.StaticValueProvider.of(region));
+    }
+
+    public DynamoDBConfiguration withRegion(ValueProvider<String> region) {
+      return builder().setRegion(region).build();
+    }
+
+    public DynamoDBConfiguration withEndpointUrl(String endpointUrl) {
+      return withEndpointUrl(ValueProvider.StaticValueProvider.of(endpointUrl));
+    }
+
+    public DynamoDBConfiguration withEndpointUrl(ValueProvider<String> endpointUrl) {
+      return builder().setRegion(endpointUrl).build();
+    }
+
+    public DynamoDBConfiguration withAwsAccessKey(String accessKey) {
+      return withAwsAccessKey(ValueProvider.StaticValueProvider.of(accessKey));
+    }
+
+    public DynamoDBConfiguration withAwsAccessKey(ValueProvider<String> accessKey) {
+      return builder().setAwsAccessKey(accessKey).build();
+    }
+
+    public DynamoDBConfiguration withAwsSecretKey(String secretKey) {
+      return withAwsSecretKey(ValueProvider.StaticValueProvider.of(secretKey));
+    }
+
+    public DynamoDBConfiguration withAwsSecretKey(ValueProvider<String> secretKey) {
+      return builder().setAwsSecretKey(secretKey).build();
+    }
+
+    AmazonDynamoDB buildAmazonDynamoDB() {
+      AmazonDynamoDBClientBuilder builder = AmazonDynamoDBClientBuilder.standard();
+      if (getEndpointUrl() != null) {
+        builder.setEndpointConfiguration(
+            new AwsClientBuilder.EndpointConfiguration(getEndpointUrl().get(), getRegion().get()));
+      }
+      if (getEndpointUrl() == null && getRegion() != null) {
+        builder.setRegion(getRegion().get());
+      }
+      if (getAwsAccessKey() != null && getAwsSecretKey() != null) {
+        builder.setCredentials(
+            new AWSStaticCredentialsProvider(
+                new BasicAWSCredentials(getAwsAccessKey().get(), getAwsSecretKey().get())));
+      }
+      return builder.build();
+    }
+  }
+
+  /** Read data from DynamoDB and return PCollection<Map<String, AttributeValue>>. */
+  @AutoValue
+  public abstract static class Read
+      extends PTransform<PBegin, PCollection<Map<String, AttributeValue>>> {
```

Comment:
Any reason to not return here better `ScanResult` ? we can if needed provide an extra transformation to extract those into `Map<String, AttributeValue>` but the more generic we can let the API the better.

Reply:
This is to make it consistent with other IO e.i. Kinesis, SNS, SQS, ... They are all return one item/record at a time. In Dynamo, a Map represent for just one record, which its content is a list of attributes/columns. Let me know if we still want ScanResult, instead?

Reply:
Yes, it is probably better to let `ScanResult` so please change it, that way users can decide what to do with it and we will solve issues of missing features if Amazon decides to add more stuff to it which we will have to maintain otherwise.

Reply:
@iemejia , I understand your point, and I am trying to make it return ScanResult. But, this is not an easy change at all, and seem to not align with the existing IOs model. First, we have to `extends PTransform<PBegin, PCollection<ScanResult>>` , and the ScanResult already contain a list of item. This is like a collection contains a collection, but the outsider one always have one item (note: can't do `extends PTransform<PBegin, ScanResult>` directly without PCollection). Second, if we  return ScanResult, what would we do with the Reader's getCurrent() and get advance()? since they are supposed to return one item at time from ScanResult? Any idea of how it would be done with ScanResult?

Reply:
@iemejia I think we are a little confused about the request to return ScanResult b/c that type represents a collection of results so it would seem natural to "unroll" the results and pass them on to downstream transforms for individual processing. If the goal is to allow more control on how consumers handle ScanResults, what do you think about changing DynamoDBIO API to include something similar to JdbcIO's RowMapper class, but with signature f: ScanResult -> [T]?

Reply:
Maybe this not too different from what you were referring to here @iemejia?
> we can if needed provide an extra transformation to extract those into Map<String, AttributeValue>

Reply:
Yes that's exactly the idea @randomsamples. I am trying to do the default to be as generic and  closer to the Amazon API as possible. That way the users can decide what to do with such objects afterwards which guarantees (1) less maintenance from our side since woe already cover the generic case and (2) that we don't need to change our API if Amazon adds something new to their default Result object to support it.

Reply:
@cmachgodaddyThe IO model is to output elements from requests, in many cases these are selfcontained records where the mapping is done 1 to 1, but there are already IOs that follow other  patterns e.g. `HBaseIO` does the same and it returns a `Result` that contains column families and columns that the user can decide to iterate in further transforms and produce more outputs (user can decide if it uses it more as a FlatMap than as a Map).

> note: can't do extends PTransform<PBegin, ScanResult> directly without Pcollection.

Yes you are right, maybe I was not clear in my first message. I intended to say `extends PTransform<PBegin, PCollection<ScanResult>>`

> Second, if we return ScanResult, what would we do with the Reader's getCurrent() and getAdvance()? since they are supposed to return one item at time from ScanResult? Any idea of how it would be done with ScanResult?

We can just add a facility  to extract the values, but we may not need to, the user can do that in its own DoFn logic.

One extra advantage of this pattern is that it will allow us to do in the future a transform of the ReadAll kind, e.g. `PTransform<PCollection<ScanRequest>, Pcollection<ScanResult>>` and with this we may do queries to multiples tables in the same pipeline.

Reply:
Thanks for your further explanation, @iemejia , I am trying this pattern with pointers that @randomsamples shared above.

Reply:
Hi @iemejia and @randomsamples, I have a closer look at HBaseIO, JdbcIO, and even CassandraIO while trying to refactor this DynamoIO base on this comment, to return ScanResult directly to user, but then realized we can't simply just return ScanResult. Here are the reasons and what I found: 
1. HBaseIO's Reader return Result, which represent for **one** row, but ScanResult in Dynamo represent for a list of row, `ScanResult.getItems()` , which returns List<Map<String, AttributeValue>>
Ref to HBase Reader: https://github.com/apache/beam/blob/master/sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java#L460
2. JdbcIO's Reader use RowMapper to send result type of T to user, but each "result" it send is one row, not a whole set of row or a list of row. (While ScanResult in Dynamo is a set of rows, again) https://github.com/apache/beam/blob/master/sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcTestHelper.java#L34 . This pointer is more clear, which create a row from the "result" object sent from Reader.
3. Similar for Cassandra....
With that being said, the ScanResult is incomparable with other "result" objects sent from other Readers to user. Second, all of those Readers iterate through each row and send to user, so if in DynamoIO, we send the ScanResult object to user directly, we don't iterate through each row, and which we will lose the ability to do parallel scan in Dynamo and split in Beam.
Hope I explained it well this one. And please let me know if I missed anything? or if you have any idea?

Reply:
@iemejia ?

Reply:
Sorry for late answer. I am trying to understand why this is different, given that each partition is being queried by one individual `ScanRequest` that assigns an independent segment, so each read of a 'partition' ends up being the responsibility of an individual worker, so doing this won't affect parallelism, and make the mapping more straight forward if we want to move it in the future to a ParDo based implementation like those IOs that use the `ReadAll` approach.

I know that this sounds contra intuitive because of the 1 to 1 mapping instead of 1 to n but it really does not make a big difference.

Only possible issue I see is that if there is an unbalanced partition it can make the read slow, but this will be the case if we 'extract' the data from the object so in the end it will be similar..

Reply:
Sorry for late answer. I am trying to understand why this is different, given that each partition is being queried by one individual `ScanRequest` that assigns an independent segment, so each read of a 'partition' ends up being the responsibility of an individual worker, so doing this won't affect parallelism, and make the mapping more straight forward if we want to move it in the future to a ParDo based implementation like those IOs that use the `ReadAll` approach.

I know that this sounds contra intuitive because of the 1 to 1 mapping instead of 1 to n but it really does not make a big difference.

Only possible issue I see is that if there is an unbalanced partition it can make the read slow, but this will be the case if we 'extract' the data from the object so in the end it will be similar..

Reply:
Thanks for the response again, @iemejia . I good to go with the ScanResult approach, just can't picture how client/user will consume it? I have `stage` implementation of it in my internal repo PR, https://github.com/cmachgodaddy/beam/pull/1/files#diff-127519743356d99d2d217126be4f240eR233
. But again I'm not quite sure know how user is going to implement the RowMapper interface and return one item or a list of item?
https://github.com/cmachgodaddy/beam/pull/1/files#diff-392c3e0413082935785b6e8f8864a3e2R179
I am getting stuck right here actually in my test case, https://github.com/cmachgodaddy/beam/pull/1/files#diff-fc5ff3bb6a92b5649de50d3ef3fb5a13R74 
Look forward to your advice?

Reply:
@iemejia , I have tried several times to implement the Reader to return ScanResult with RowMapper (like JDBC's one, as suggested above), but still couldn't get the unit test work. Would you mind to take a look at my repo?

Here is the RowMapper: https://github.com/cmachgodaddy/beam/blob/ScanResultApproach/sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java#L119

And this is my ReadFn which call the RowMapper and pass the ScanResult to user:
https://github.com/cmachgodaddy/beam/blob/ScanResultApproach/sdks/java/io/amazon-web-services/src/main/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIO.java#L214

And, here is the implementation of the RowMapper from user/client side:
https://github.com/cmachgodaddy/beam/blob/ScanResultApproach/sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTestHelper.java#L177

The question is should user/client return ScanResult directly or a List<Items> which attracted from ScanResult? Neither ways, unit testing both cases are getting stuck.
https://github.com/cmachgodaddy/beam/blob/ScanResultApproach/sdks/java/io/amazon-web-services/src/test/java/org/apache/beam/sdk/io/aws/dynamodb/DynamoDBIOTest.java#L74

I am not sure if you understand the issue I am dealing with this ScanResult approach? I would love to jump on a call to explain more and answer any questions that you have. Please let me know when you're available for a chat? Or if you have any idea to speed up this PR since it's has been out there for more than a couple of months?

Reply:
Yes, I  agree with you, this PR has taken way too long. Please excuse me if my delayed review has contributed to that. However I am a bit busy this weekend, so I will take a look and give you feedback on monday.

The next release of Beam (2.14.0) will be the 19/06 so if you want we can try to finish it before that date. I will mark the Jira to be included in next release and we’ll see if we can make it.

If you prefer to discuss more interactively you can ping me at the [ASF slack](http://the-asf.slack.com/) You can [request an invite here](https://s.apache.org/slack-invite). Notice that I am based in Europe so consider the timezone difference for this.

Reply:
@iemejia , let try to get it in 19/06. Can you review my last commit? and the issue that I have with ScanResult in my repo?

Reply:
Is there any technical reason that impeaches returning `ScanResult`? I already explained above my reasons to prefer this, summarizing:

1. I prefer to keep the AWS object available (in case they expand their API).
2. Parallelism is not affected because the user will have one ScanResult per worker so he can just extract and iterate the values.

I understand that (2) may seem less user-friendly, but it is in order to be consistent with (1). It is mostly about being 'independent' of the evolution of the Amazon Dynamo API and of future requests from users around things not available because of not exposing the `ScanResult` API.

Reply:
@iemejia , of course, I love to go with your approach but, yes there is a technical channel to implement the Reader with ScanResult. Would you  mind to look at my explanation and my repo above, for more details?